update ds/binary-heap (#33)
OI-wiki/i18n: docs/ds/binary-heap.md

auhor: HeRaNO, Xeonacid

### Structure

For the structure of the binary heap, it is a binary tree, and a complete binary tree, where each node stores an element (or a weight).

Property of heap: the weight of parent node is not less than the weight of child node (max heap). Similarly, we can define a min heap. This article takes the max heap as an example. 

Due to the nature of the heap, the root of the tree stores the maximum value (which can be obtained using getmax operation).

### Insertion operation

The insertion operation refers to inserting an element into the binary heap, and it is required to still be a complete binary tree after insertion.

The easiest way is to insert after the rightmost leaf in the bottom layer.

If the bottom layer is full, then add another layer.

> If cannot satisfy the property of heap after insertion?

Reviewer: ```suggestion
How about the property of heap cannot be satisfied after insertion?
```

 **Upward adjustment**: If the weight of this node is greater than the weight of its parent node, swap and repeat this process until this condition is no longer satisfied or reaches the root.

It can be proved that after the upward adjustment in insertion operation, no other node will not satisfy the heap property.

The time complexity of the upward adjustment is $O(\log n)$ .

### Deletion operation

Deletion operation refers to deleting the largest element in the heap, that is, deleting the root node.

But if deleted directly, there will be two heaps, which is difficult to handle.

> So consider the reverse process of the insertion operation, try to move the root to the last node, and then delete it directly.

Reviewer: ```suggestion
So consider the reverse process of the insertion operation. That is, try to move the root to the last node, and then delete it directly.
```

However, it is difficult to do in practice. The method we usually use is to directly swap the root node with the last node.

So the root node (at the last node) is deleted directly, but the new root node may not satisfy the heap property...

 **Downward adjustment**: Find the largest one among the child nodes, swap it with the root node, and repeat this process to the bottom.

It can be proved that after deletion and downward adjustment, no other node does not satisfy the heap property.

Time complexity is $O(\log n)$ .

### Decrease the weight of a node

Obviously, after direct changes, you can adjust it upward once, and the time complexity is $O(\log n)$ .

### Implementation

> We found that several operations described above mainly rely on two principles: upward adjustment and downward adjustment.

Reviewer: concise: upward and downward adj.s

> Consider using a sequence $h$ to represent the heap. The two child nodes of $h_i$ are $h_{2i}$ and $h_{2i+1}$ respectively, and $1$ is the root node:

Reviewer: ```suggestion
Consider using a sequence $h$ to represent the heap. The two child nodes of $h_i$ are $h_{2i}$ and $h_{2i+1}$, respectively, and $1$ is the root node:
```

![heap structure of h](./images/binary-heap1.png)

> Code sample:

Reviewer: Sample code

### Implementation

> We found that several operations described above mainly rely on two principles: upward and downward adjustment.

Reviewer: 昨天suggest的时候忘记最后改复数了 qwq

Author: 我的锅..忘记复数加s了 🤕

Consider using a sequence $h$ to represent the heap. The two child nodes of $h_i$ are $h_{2i}$ and $h_{2i+1}$ , respectively, and $1$ is the root node:

![heap structure of h](./images/binary-heap1.png)

Sample code:

```cpp

> void up(inst xs) {

Reviewer: typo

}
```

### Building a heap

Consider this problem: starting from an empty heap, insert $n$ elements regardless of its order.

Inserting directly one by one takes $O(n \log n)$ time. Is there a better way?

> #### Method 1: Use decreasekey (ie, upward adjustment)

Reviewer: i. e. 
注意空格（

Inserting directly one by one takes $O(n \log n)$ time. Is there a better way?

> #### Method 1: Use decreasekey(i.e. upward adjustment)

Reviewer: ```suggestion
#### Method 1: Use decreasekey (i.e. upward adjustment)
```
